1\. **Pandas DataFrame**

This exercise consists in analyzing a dataset containg timing information from a series of Time-to-Digital-Converters (TDC) implemented in a pair of FPGAs. Each measurement (i.e. each row of the input file) consists of a flag that specifies the type of message ('HEAD', which in this case is always 1), two addresses of the TDC providing the signal ('FPGA' and 'TDC_CHANNEL'), and the timing information ('ORBIT_CNT', 'BX_COUNTER', and 'TDC_MEAS'). Each TDC count corresponds to 25/30 ns, whereas a unit of BX_COUNTER corresponds to 25 ns, and the ORBIT_CNT is increased every `x` BX_COUNTER. This allows to store the time in a similar way to hours, minutes and seconds.

In [4]:
# If haven't downloaded it yet, please get the data file with wget
#!wget https://www.dropbox.com/s/xvjzaxzz3ysphme/data_000637.txt -P ./data/

1\. Create a Pandas DataFrame reading N rows of the `data/data_000637.txt` dataset. Choose N to be smaller than or equal to the maximum number of rows and larger that 10k (check the documentation).

In [27]:
import pandas as pd
import random as rnd

file_name = "./data/data_000637.txt"
data = pd.read_csv(file_name) #leggo il file

#creo un numero random tra 10000 e il numero massimo di righe
N=rnd.randint(10000,len(data))
print(f"N: {N}")

print(f"Total numbers of rows: {len(data)}")

#seleziono e stampo le prime N righe
res = data.head(N)
res

N: 984525
Total numbers of rows: 1310720


HEAD  FPGA  TDC_CHANNEL   ORBIT_CNT  BX_COUNTER  TDC_MEAS
0          1     0          123  3869200167        2374        26
1          1     0          124  3869200167        2374        27
2          1     0           63  3869200167        2553        28
3          1     0           64  3869200167        2558        19
4          1     0           64  3869200167        2760        25
...      ...   ...          ...         ...         ...       ...
984520     1     0           76  3869208915        3316         6
984521     1     0           49  3869208915        3488        18
984522     1     0          139  3869208915        3495         0
984523     1     0           56  3869208915        3488         0
984524     1     0           63  3869208915        3493        29

[984525 rows x 6 columns]

2\. Estimate the number of BX in a ORBIT (the value `x`).

*Hint*: check when the BX counter reaches the maximum value before being reset to 0.

In [78]:
tmp=data.loc[[0],'BX_COUNTER'].values[0]
print(tmp)
counter=0
numb=[]
for i in range(128, len(data)):
    tmp2=data.loc[[i+1],'BX_COUNTER'].values[0]
    tmp1=data.loc[[i],'BX_COUNTER'].values[0]
    if(tmp2-tmp1<-20):
        break
    else:
        counter=counter+1

print(counter+1)

2374
127


3\. Create a new column with the absolute time in ns (as a combination of the other three columns with timing information) since the beginning of the data acquisition, and convert the new column to a Time Series.

4\. Find out the duration of the data taking in hours, minutes and seconds, by using the features of the Time Series. Perform this check reading the whole dataset.

5\. Use the `.groupby()` method to find out the noisy channels, i.e. the TDC channels with most counts (print to screen the top 3 and the corresponding counts)

6\. Count the number of non-empty orbits (i.e. the number of orbits with at least one hit).

7\. Count the number of unique orbits with at least one measurement from TDC_CHANNEL=139.

8\. Create two Series (one for each FPGA) that have the TDC channel as index, and the number of counts for the corresponding TDC channel as values.

9\. **Optional:** Create two histograms (one for each FPGA) that show the number of counts for each TDC channel.